Added private repo access error catch and missing repo error capture

sachin-sandhu opened the pull request:
### What are you trying to accomplish?

<!-- Provide both a what and a _why_ for the change. -->

Fixes following sentry issue : 

```
NoMethodError: undefined method `requirements' for nil:NilClass

          sources = dependency.requirements
                              ^^^^^^^^^^^^^
  npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:298:in `registry_source_url'
  npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:137:in `locked_registry'
  npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:34:in `registry'
  npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb:130:in `raise_package_access_error'
  npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb:108:in `block in handle_pnpm_lock_updater_error'
...
(48 additional frame(s) were not displayed)

```
<!-- What issues does this affect or fix? -->




**Preface :**  Dependencies are listed in public as well as private registry sources. When a dependency is listed in private registry, Information and authentication to the private registry is usually defined in .npmrc file or configured locally using NPM CLI. In case :dependabot: is unable to access the package because of no information regarding private registry (i.e. no .npmrc file present in repo), native helper will throw a **ERR_PNPM_FETCH_404** . i.e.

`ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@gsap%2Fbusiness:  - 404`

This error is filtered by :dependabot: native helper and an appropriate exception `PrivateSourceAuthenticationFailure` is thrown for the dependency. 

**Issues :** 
1. The message returned from NPM and PNPM native helper is returned with slight difference compared to one returned from :dependabot: job run. On running the update using _PNPM CLI_, the error message returned is as following:

`ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@gsap%2Fbusiness: Not Found - 404`

while on running the update from :dependabot: job, the message returned is as following:

`ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@gsap%2Fbusiness:  - 404` 

**Fix :** Matching with qualifying error description should capture the exception handling. So added a new error description match 

`MISSING_PACKAGE_IN_REPO = /ERR_PNPM_FETCH_404[ [^:print:]]+GET (?<dependency_url>.*):  - 404/`

2. Currently in `raise_package_access_error` (in npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb), used to raise PrivateSourceAuthenticationFailure exception, the context variable `missing_dep` is used to capture the context information (missing dependency) for  exception `PrivateSourceAuthenticationFailure` . However when lock files (package-lock.json , pnpm-lock.yaml) have listed dependencies source in format "gsap": "**npm**:@gsap/business@^3.12.4"  the context variable is returned as nil. This is likely to be issue with dependency parser _NpmAndYarn::FileParser_. Due to this, sources = **dependency**.requirements return exception NoMethodError: undefined method `requirements for nil:NilClass` (main error) . To fix this issue, `missing_dep` is checked for nil and exception `PrivateSourceAuthenticationFailure` is raised with context variable`package_name` that contains the missing package name. the exception handled looks like this
`Handled error whilst updating next: private_source_authentication_failure "@gsap/business"`. This should fix the NoMethodError unhandled exception and instead raise PrivateSourceAuthenticationFailure exception.

 

### Anything you want to highlight for special attention from reviewers?

<!-- If there were multiple ways to approach the problem, why did you pick this one? -->

### How will you know you've accomplished your goal?

`NoMethodError: undefined method `requirements' for nil:NilClass` error should be halted in Sentry.

<!--
  * If you've reproduced an error, can you link to, or demonstrate the reproduction?
  * If you've added a new feature, how will you demonstrate it to others?
  * If you've refactored code, how will you demonstrate that the new code is functionally equivalent to the old code?
-->

### Checklist

<!-- Before requesting review, please ensure that your pull request fulfills the following requirements: -->

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

honeyankit commented on npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb:
```diff
@@ -43,6 +43,7 @@ def updated_pnpm_lock_content(pnpm_lock)
         FORBIDDEN_PACKAGE = /ERR_PNPM_FETCH_403[ [^:print:]]+GET (?<dependency_url>.*): Forbidden - 403/
         MISSING_PACKAGE = /ERR_PNPM_FETCH_404[ [^:print:]]+GET (?<dependency_url>.*): Not Found - 404/
```
Since the error output differs for just `Not Found`, can we optimize the `MISSING_PACKAGE` itself. Something like below:

```bash
  MISSING_PACKAGE = /ERR_PNPM_FETCH_404[ [^:print:]]+GET (?<dependency_url>.*):(?: Not Found)? - 404/
```

sachin-sandhu replied:
@honeyankit , i have fixed the string reg ex, thanks for the suggestion.

honeyankit reviewed (commented):
Since the error output differs for just `Not Found`, can we optimize the `MISSING_PACKAGE` itself. Something like below in which `Not Found` is made optional for missing package check.

sachin-sandhu commented:
> @sachin-sandhu Could we add test cases for both the changes you are doing. That way, we can be sure that the we have fixed the issue.

@honeyankit , i have added added test cases for change validation.

honeyankit commented on common/lib/dependabot/errors.rb:
```diff
@@ -191,6 +191,11 @@ def self.updater_error_details(error)
         "error-type": "private_source_authentication_failure",
         "error-detail": { source: error.source }
       }
+    when Dependabot::MissingDependencyInRegistry
```
```suggestion
    when Dependabot::DependencyNotFound
```

honeyankit replied:
The package is not missing from the registry, the package is missing from the lockfile, so suggesting better name.

honeyankit commented on npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater_spec.rb:
```diff
@@ -195,5 +195,34 @@
+
+      it "raises a helpful error" do
+        expect { updated_pnpm_lock_content }
+          .to raise_error(Dependabot::MissingDependencyInRegistry)
```
```suggestion
          .to raise_error(Dependabot:: DependencyNotFound)
```

honeyankit commented on npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb:
```diff
@@ -123,6 +123,8 @@ def raise_package_access_error(dependency_url, pnpm_lock)
           missing_dep = lockfile_dependencies(pnpm_lock)
                         .find { |dep| dep.name == package_name }
 
+          raise MissingDependencyInRegistry, package_name unless missing_dep
```
```suggestion
          raise DependencyNotFound, package_name unless missing_dep
```

honeyankit reviewed (commented):
Suggested better name for raising the new error
